File: backend/app/api/v1/jobs.py
"""Job status and enqueue endpoints for background work (Phase 9)."""

from __future__ import annotations

from fastapi import APIRouter, status

from app.core.exceptions import NotFoundError, ValidationAppError
from app.schemas.common import DataResponse
from app.schemas.job import (
    EmailEnqueueRequest,
    JobEnqueueResult,
    JobStatusRead,
    ReportGenerateRequest,
)
from app.services.jobs import JobService

router = APIRouter(tags=["jobs"])


@router.get(
    "/jobs/{job_id}",
    response_model=DataResponse[JobStatusRead],
)
def get_job_status(job_id: str) -> DataResponse[JobStatusRead]:
    """Poll RQ job status. Never blocks on job execution."""
    service = JobService()
    if not service.redis_available():
        raise ValidationAppError(
            "Background job service is unavailable.",
            fields=[{"field": "redis", "message": "Redis is not reachable."}],
        )
    payload = service.get_job_status(job_id)
    if payload is None:
        raise NotFoundError("JOB_NOT_FOUND", "Job not found.")
    return DataResponse(data=JobStatusRead.model_validate(payload))


@router.post(
    "/reports/generate",
    response_model=DataResponse[JobEnqueueResult],
    status_code=status.HTTP_202_ACCEPTED,
)
def generate_report(
    payload: ReportGenerateRequest,
) -> DataResponse[JobEnqueueResult]:
    """Enqueue report generation and return immediately with a job id."""
    if payload.report_type == "project_summary" and payload.project_id is None:
        raise ValidationAppError(
            "project_id is required for project_summary reports.",
            fields=[{"field": "project_id", "message": "Required."}],
        )

    service = JobService()
    job_id = service.enqueue_report(
        report_type=payload.report_type,
        project_id=payload.project_id,
    )
    if job_id is None:
        raise ValidationAppError(
            "Could not enqueue report job. Is Redis running?",
            fields=[{"field": "redis", "message": "Enqueue failed."}],
        )
    return DataResponse(
        data=JobEnqueueResult(job_id=job_id, status="queued"),
        message="Report generation queued.",
    )


@router.post(
    "/jobs/email",
    response_model=DataResponse[JobEnqueueResult],
    status_code=status.HTTP_202_ACCEPTED,
)
def enqueue_email(
    payload: EmailEnqueueRequest,
) -> DataResponse[JobEnqueueResult]:
    """Enqueue an email job (infrastructure path for Phase 9 email workers)."""
    service = JobService()
    job_id = service.enqueue_email(
        to=payload.to,
        subject=payload.subject,
        body=payload.body,
    )
    if job_id is None:
        raise ValidationAppError(
            "Could not enqueue email job. Is Redis running?",
            fields=[{"field": "redis", "message": "Enqueue failed."}],
        )
    return DataResponse(
        data=JobEnqueueResult(job_id=job_id, status="queued"),
        message="Email job queued.",
    )

File: frontend/services/jobs.ts
/**
 * Background job client (Phase 9).
 * Enqueues report/email work and polls job status — never waits on workers.
 */

import type {
  ApiSuccess,
  JobEnqueueResult,
  JobStatus,
} from "@/types/domain";

import { isoNow, mockForceError, newId } from "@/services/mock/data";
import { delay } from "@/services/mock/delay";

const USE_MOCK = true;

const mockJobs = new Map<
  string,
  JobStatus & { created_at: string; report_type?: string }
>();

export async function generateReport(input: {
  report_type?: string;
  project_id: string;
}): Promise<ApiSuccess<JobEnqueueResult>> {
  if (USE_MOCK) {
    await delay(120);
    if (mockForceError) throw new Error("Failed to enqueue report.");
    const jobId = newId("job");
    const stamp = isoNow();
    mockJobs.set(jobId, {
      id: jobId,
      status: "finished",
      description: `generate_report:${input.report_type ?? "project_summary"}`,
      enqueued_at: stamp,
      started_at: stamp,
      ended_at: stamp,
      result: {
        status: "ok",
        report_type: input.report_type ?? "project_summary",
        project_id: input.project_id,
        summary: {
          asset_count: 0,
          document_count: 0,
          assets_by_status: {},
          assets_by_type: {},
        },
      },
      error: null,
      created_at: stamp,
      report_type: input.report_type ?? "project_summary",
    });
    return {
      success: true,
      data: { job_id: jobId, status: "queued" },
      message: "Report generation queued.",
    };
  }
  throw new Error("Live API not enabled");
}

export async function getJobStatus(
  jobId: string,
): Promise<ApiSuccess<JobStatus>> {
  if (USE_MOCK) {
    await delay(80);
    const job = mockJobs.get(jobId);
    if (!job) throw new Error("Job not found.");
    return {
      success: true,
      data: {
        id: job.id,
        status: job.status,
        description: job.description,
        enqueued_at: job.enqueued_at,
        started_at: job.started_at,
        ended_at: job.ended_at,
        result: job.result,
        error: job.error,
      },
      message: null,
    };
  }
  throw new Error("Live API not enabled");
}
